agent/tracer.go

```diff
@@ -1,31 +1,140 @@
 package agent
 
 import (
+	"context"
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
 	"net/url"
+	"sync"
 	"time"
 
+	"github.com/elastic/hey-apm/conv"
+	"github.com/elastic/hey-apm/strcoll"
+
 	"go.elastic.co/apm"
 	apmtransport "go.elastic.co/apm/transport"
 )
 
-func Tracer(logger apm.Logger, serverUrl, serverSecret string, maxSpans int) *apm.Tracer {
-	tracer := apm.DefaultTracer
-	tracer.SetLogger(logger)
-	tracer.SetMetricsInterval(0) // disable metrics
-	transport := tracer.Transport.(*apmtransport.HTTPTransport)
-	transport.SetUserAgent("hey-apm")
+type Tracer struct {
+	*apm.Tracer
+	TransportStats *transportStats
+}
+
+type transportStats struct {
+	Accepted  float64
+	TopErrors []string
+}
+
+func (t Tracer) Close() {
+	t.Tracer.Close()
+	t.Transport.(*transport).wg.Wait()
+	close(t.Transport.(*transport).responses)
+}
+
+func NewTracer(logger apm.Logger, serverUrl, serverSecret string, maxSpans int) *Tracer {
+
+	goTracer := apm.DefaultTracer
+	transport := wrap(goTracer.Transport.(*apmtransport.HTTPTransport), serverUrl, serverSecret)
+	goTracer.Transport = transport
+	goTracer.SetLogger(logger)
+	goTracer.SetMetricsInterval(0) // disable metrics
+	goTracer.SetSpanFramesMinDuration(1 * time.Nanosecond)
+	goTracer.SetMaxSpans(maxSpans)
+
+	tracer := &Tracer{goTracer, &transportStats{}}
+
+	go func() {
+		for {
+			select {
+			case response := <-transport.responses:
+				var m map[string]interface{}
+				if err := json.Unmarshal(response, &m); err != nil {
+					return
+				}
+				tracer.TransportStats.Accepted += conv.AsFloat64(m, "accepted")
+				for _, i := range conv.AsSlice(m, "errors") {
+					e := conv.AsString(i, "message")
+					if !strcoll.Contains(e, tracer.TransportStats.TopErrors) {
+						tracer.TransportStats.TopErrors = append(tracer.TransportStats.TopErrors, e)
+					}
+				}
+				transport.wg.Done()
+			}
+		}
+	}()
+	return tracer
+}
+
+type transport struct {
+	*apmtransport.HTTPTransport
+	headers   http.Header
+	url       *url.URL
+	responses chan []byte
+	wg        sync.WaitGroup
+}
+
+func wrap(backend *apmtransport.HTTPTransport, serverUrl, serverSecret string) *transport {
```

Comment:
As mentioned in https://github.com/elastic/hey-apm/pull/100#discussion_r274889397, I was hoping we could avoid duplicating the internals of the go agent since the goal is just to access the req/rsp to apm-server.  Please correct me if I've missed something else we need.

I took a stab at it since I wasn't certain it's possible, this seems to do the job:

```patch
diff --git a/main.go b/main.go
index a4930c6..1f8ddf2 100644
--- a/main.go
+++ b/main.go
@@ -1,10 +1,14 @@
 package main
 
 import (
+	"bytes"
 	"flag"
+	"fmt"
+	"io/ioutil"
 	"log"
 	"math"
 	"math/rand"
+	"net/http"
 	"net/url"
 	"os"
 	"time"
@@ -13,6 +17,38 @@ import (
 	apmtransport "go.elastic.co/apm/transport"
 )
 
+type rt struct{}
+
+func (r *rt) RoundTrip(req *http.Request) (*http.Response, error) {
+	if req.URL.Path != "/intake/v2/events" {
+		return http.DefaultTransport.RoundTrip(req)
+	}
+
+	fmt.Println("making intake call")
+	q := req.URL.Query()
+	q.Set("verbose", "")
+	req.URL.RawQuery = q.Encode()
+
+	rsp, err := http.DefaultTransport.RoundTrip(req)
+	if err != nil {
+		return rsp, err
+	}
+
+	if rsp.Body != http.NoBody && rsp.StatusCode == http.StatusAccepted {
+		var buf bytes.Buffer
+		if _, err = buf.ReadFrom(rsp.Body); err != nil {
+			return rsp, err
+		}
+		if err = rsp.Body.Close(); err != nil {
+			return rsp, err
+		}
+		fmt.Println(buf.String())
+
+		rsp.Body = ioutil.NopCloser(&buf)
+	}
+	return rsp, err
+}
+
 func main() {
 	// run options
 	runTimeout := flag.Duration("run", 30*time.Second, "stop run after this duration")
@@ -80,6 +116,7 @@ func main() {
 		}
 		transport.SetServerURL(u)
 	}
+	transport.Client.Transport = &rt{}
 
 	tracer.SetSpanFramesMinDuration(1 * time.Nanosecond)
 	tracer.SetMaxSpans(*spanMaxLimit)
```

producing:
```
hey-apm -run 3s
2019/04/25 17:34:14 log.go:12: [debug] random seed: 1556228054
2019/04/25 17:34:14 log.go:12: [debug] start
making intake call
{"accepted":6410}

2019/04/25 17:34:15 log.go:12: [debug] sent request with 443 transactions, 3405 spans, 2562 errors, 0 metricsets
making intake call
2019/04/25 17:34:17 main.go:138: generated 774287 events (errors: 277010, transctions: 76631, spans: 420646) in 3.001126625s
{"accepted":7907}
...
```

What do you think?

Reply:
great idea, didn't know you could do that. copy-pasted that here c97a6f369fafe62b19fbe4c0d9e1be529c61dd7c